services/microsoft.network/resource.ftl

```diff
@@ -0,0 +1,523 @@
+the standalone Resource, whereas the Object function defines the itterable 
+property object. This allows flexability for Component authors.
+--]
+
+[#assign VIRTUAL_NETWORK_OUTPUT_MAPPINGS = 
+  {
+    REFERENCE_ATTRIBUTE_TYPE : {
+      "Property" : "id"
+    }
+  }
+]
+
+[#assign SUBNET_OUTPUT_MAPPINGS =
+  {
+    REFERENCE_ATTRIBUTE_TYPE : {
+      "Property : "id"
+    }
+  }
+]
+
+[#assign outputMappings += 
+  {
+    AZURE_VIRTUAL_NETWORK_RESOURCE_TYPE : VIRTUAL_NETWORK_OUTPUT_MAPPINGS,
+    AZURE_SUBNET_RESOURCE_TYPE : SUBNET_OUTPUT_MAPPINGS
+  }
+]
+
+[#function getApplicationSecurityGroupObject 
+  resourceId
+  location
+  tags={}
+  properties={}]
+
+  [#return 
+    {
+      "id" : resourceId,
+      "location" : location,
+      "properties" : properties
+    } +
+    attributeIfContent("tags", tags)
+  ]
+[/#function]
+[#macro createApplicationSecurityGroup name location tags={}]
+  [@armResource
+    name=name
+    type="Microsoft.Network/applicationSecurityGroups"
+    apiVersion="2019-04-01"
+    location=location
+    tags=tags
+  /]
+[/#macro]
+
+[#function getNetworkSecurityGroupSecurityRulesObject
+  protocol
+  access
+  direction
+  sourceAddressPrefix=""
+  sourceAddressPrefixes=[]
+  sourcePortRange=""
+  sourcePortRanges=[]
+  sourceApplicationSecurityGroups=[]
+  destinationPortRange=""
+  destinationPortRanges=[]
+  destinationAddressPrefix=""
+  destinationAddressPrefixes=""
+  destinationApplicationSecurityGroups=[]
+  description=""
+  priority=""]
+
+  [#return
+    {
+      "access" : access,
+      "direction" : direction,
+      "protocol" : protocol
+    } +
+    attributeIfContent("sourceAddressPrefix", sourceAddressPrefix) +
+    attributeIfContent("sourceAddressPrefixes", asArray(sourceAddressPrefixes)) +
+    attributeIfContent("sourcePortRange", sourcePortRange) +
+    attributeIfContent("sourcePortRanges, asArray(sourcePortRanges)) +
+    attributeIfContent("sourceApplicationSecurityGroups", asArray(sourceApplicationSecurityGroups)) +
+    attributeIfContent("destinationPortRange", destinationPortRange) +
+    attributeIfContent("destinationPortRanges", asArray(destinationPortRanges)) +
+    attributeIfContent("destinationAddressPrefix", destinationAddressPrefix) +
+    attributeIfContent("destinationAddressPrefixes", asArray(destinationAddressPrefixes)) +
+    attributeIfContent("destinationApplicationSecurityGroups", asArray(destinationApplicationSecurityGroups)) +
+    attributeIfContent("description", description) +
+    attributeIfContent("priority", priority)
+  ]
+
+[/#function]
+[#macro createNetworkSecurityGroupSecurityRules
+  name
+  properties
+  tags={}
+  outputs={}
+  dependsOn=[]]
+
+  [@armResource
+    name=name
+    type="Microsoft.Network/networkSecurityGroups/securityRules"
+    apiVersion="2019-04-01"
+    dependsOn=dependsOn
+    properties=properties
+    tags=tags
+    outputs=outputs
+  /]
+  
+[/#macro]
+
+[#function getRouteTableRouteObject 
```

Comment:
So to create the routetable you would need to call getRouteTableRouteObject then pass it to the createRouteTableRoute macro as the properties value?

Reply:
If there isn't a limit to the number of resources or some weird functionality in the dedicated resource vs the combined resource I'd be tempted to go with using the separate resources all the time. This would still provide flexibility to component developers while also reducing the number of functions required.

Reply:
[There are some resource limits](https://docs.microsoft.com/en-us/azure/azure-resource-manager/template-best-practices#template-limits), though we are less likely to hit the resource limit (800) than we are to hit the max outputs of 64, if we are outputting Id values for every new resource in order to reference it. Tiers, zones, route tables + rules, NSG + rules & flowLog resources each with their own Id's alone will quickly set us on our way to that.

this approach allows for simplicity in creating arrays of these objects instead of numerous individual resources. The fully expanded VirtualNetwork resource is basically all arrays of other resources (subnets, nsgs, route tables, vnetpeerings). We don't have to combine everything into one enormous component (and I haven't) however I think that it does keep things tidy - see example below).

The total number of functions goes up, but its the same complexity. It means I'm reusing my object code in multiple places (same rules object below whether I create a full NSG with default rules as shown, if i use the macro to create a default rule resource, or a separate macro to create a standard rule resource (for some reason they are different types)). The macros become only a few lines and the object is efficiently reused.

```
    [#local nsgDefaultSecurityRules = [
      getNetworkSecurityGroupSecurityRulesObject(
        description="Deny All Outbound"
        protocol="*"
        sourcePortRange="*"
        destinationPortRange="*"
        sourceAddressPrefix="0.0.0.0/0"
        destinationAddressPrefix="0.0.0.0/0"
        access="Deny"
        priority="64998"
        direction="Outbound"
      ), 
      getNetworkSecurityGroupSecurityRulesObject(
        description="Deny All Inbound"
        protocol="*"
        sourcePortRange="*"
        destinationPortRange="*"
        sourceAddressPrefix="0.0.0.0/0"
        destinationAddressPrefix="0.0.0.0/0"
        access="Deny"
        priority="64999"
        direction="Inbound"
      )
    ]]

    [@createNetworkSecurityGroup
      name=networkSecurityGroupName
      location=regionId
      properties=
        {
          "defaultSecurityRules" : nsgDefaultSecurityRules
        }
    /]
```

Granted, in the example you referenced this is not really the case. By the point i'd written that I had found it to be a simple pattern so kept it up.

Reply:
So is it possible to still get an output of a subnet Id if a resource has been nested into another resource? Just trying to think of higher components which need to reference a particular subnet

Reply:
I'm reading your question two different ways, those being:

**Q1**. If my subnet is nested into the Vnet resource, does it still have an Id attribute which can be referenced.
**Q2**. In that same scenario, can my Vnet output the Subnet Id(s).

The answer to both is yes. 

**A1**, the Virtual Network arm resource template shows that even when nested, the Id attribute is still available to every subnet defined. This isn't for the template creator to define, but to know it can be referenced. The resource doesn't even have to exist yet. Through using the resourceId() function (made available to the plugin via PR #33 in the ``formatAzureResourceIdReference(resourceName)`` function) you will be able to reference this Id from within the same or separate template.

**A2**. If I had chosen to nest all subnets within the Vnet resource, I could then output the subnet Id's using the Output parameter of the armResource macro:
```
[#macro armResource
    name
    type
    ...
    outputs={}]

    ...

    [#list outputs as outputType,outputValue]
        [#if outputValue.UseRef!false]

            [#-- format the ARM function: resourceId() --]
            [#local reference= 
                formatAzureResourceIdReference(
                    resourceId=name
                    resourceType=type
                )
            ]

            [@armOutput
                name=name
                type="string"
                value=reference
            /]
        [#else]

            [#-- format the ARM function: reference() --] 
            [#local reference= 
                formatAzureResourceReference(
                    resourceId=name
                    resourceType=type
                    attributes=outputValue.Property
                )
            ]
       
            [@armOutput
                name=name,
                type="string",
                value=reference   
            /]

        [/#if]
    [/#list]
```

Note that this example also makes use of the new functions in PR #33.

Reply:
@roleyfoley  The [latest commit](https://github.com/codeontap/gen3-azure/pull/32/commits/781853a47bcce53569612b9be89f1438eb6af2f6) should have addressed your concerns. I came around to your way of thinking and have merged the majority of functions into the macros. I've also de-pluralised the macro names - these are COT macros and read far better when not pluralised, despite what Microsoft has called their resource types.

Once #33 is implemented I want to go back and update the armResource macro for the type property to make use of ResourceProfiles - this would mean never having to specify the apiVersion outside of MasterData which would be quite tidy.

Reply:
Thanks @RossMurr4y 
My main concern was to essentially get the Vnet and Subnet Ids into the stack outputs so that other templates can reference them. When it comes to output limits from the templates we run into the same limit in AWS. 

This is where the output mappings comes in since we can limit which outputs we actually care about. RouteTables, default NSG ids don't really need to be exposed to other components since they can't really consume them, however the subnet can be consumed by other components so we need that Id exposed